Fix: Update Rest V3 routes with correct schema implementation

defunctl opened the pull request:
<!-- Make sure to prefix the title with one of New:, Fix:, Changed:, or Security: -->

<!-- Indicate the issue(s) resolved by this PR. -->

> :rotating_light:  **NOTE:** I couldn't get tests to run locally and they seem to also fail via the GitHub actions due to outdated PHPUnit libraries, otherwise I would have added some tests for this.

## Description

<!-- Summarize the related issue, explain HOW this PR solves the problem, and WHY you made the choices you made. -->

As I started to think about the MCP Server implementation, I will need to provide it accurate JSON schemas for the responses and I noticed a few problems that this resolves:

1. In the V3 Rest controllers, the `schema` callback was incorrectly placed in the wrong spot in the array for `register_rest_route()` to actually display the schema, so it never did.
2. Most schemas were missing the `'$schema' => 'http://json-schema.org/draft-07/schema#',` property.
3. Some endpoints didn't have a schema at all, so I built those based on the responses and using AI.
4. Some endpoints schema did not match the response, and now it should.

> I ran out of time for today, but I wasn't sure where the Form REST controllers were.

Send an `OPTIONS` method request, e.g.:

```shell
curl -X https://local.lndo.site/wp-json/givewp/v3/campaigns
```

### Before

This is missing the `schema` property:

<details>

<summary>Show  JSON Response</summary>

```json
{
    "namespace": "givewp/v3",
    "methods": [
        "GET",
        "POST"
    ],
    "endpoints": [
        {
            "methods": [
                "GET"
            ],
            "args": {
                "status": {
                    "type": "array",
                    "items": {
                        "type": "string",
                        "enum": [
                            "active",
                            "draft",
                            "archived"
                        ]
                    },
                    "default": [
                        "active"
                    ],
                    "required": false
                },
                "ids": {
                    "type": "array",
                    "default": [],
                    "required": false
                },
                "page": {
                    "type": "integer",
                    "default": 1,
                    "minimum": 1,
                    "required": false
                },
                "per_page": {
                    "type": "integer",
                    "default": 30,
                    "minimum": 1,
                    "maximum": 100,
                    "required": false
                },
                "sortBy": {
                    "type": "enum",
                    "enum": [
                        "date",
                        "amount",
                        "donors",
                        "donations"
                    ],
                    "default": "date",
                    "required": false
                },
                "orderBy": {
                    "type": "enum",
                    "enum": [
                        "asc",
                        "desc"
                    ],
                    "default": "desc",
                    "required": false
                }
            }
        },
        {
            "methods": [
                "POST"
            ],
            "args": {
                "title": {
                    "type": "string",
                    "required": true
                },
                "shortDescription": {
                    "type": "string",
                    "required": false
                },
                "startDateTime": {
                    "type": "string",
                    "format": "date-time",
                    "required": false
                },
                "endDateTime": {
                    "type": "string",
                    "format": "date-time",
                    "required": false
                }
            }
        }
    ],
    "_links": {
        "self": [
            {
                "href": "https://local.lndo.site/wp-json/givewp/v3/campaigns"
            }
        ]
    }
}
```

</details>

---

### After

The schema property is now available:

<details>

<summary>Show  JSON Response</summary>

```json
{
    "namespace": "givewp/v3",
    "methods": [
        "GET",
        "POST"
    ],
    "endpoints": [
        {
            "methods": [
                "GET"
            ],
            "args": {
                "status": {
                    "type": "array",
                    "items": {
                        "type": "string",
                        "enum": [
                            "active",
                            "draft",
                            "archived"
                        ]
                    },
                    "default": [
                        "active"
                    ],
                    "required": false
                },
                "ids": {
                    "type": "array",
                    "default": [],
                    "required": false
                },
                "page": {
                    "type": "integer",
                    "default": 1,
                    "minimum": 1,
                    "required": false
                },
                "per_page": {
                    "type": "integer",
                    "default": 30,
                    "minimum": 1,
                    "maximum": 100,
                    "required": false
                },
                "sortBy": {
                    "type": "enum",
                    "enum": [
                        "date",
                        "amount",
                        "donors",
                        "donations"
                    ],
                    "default": "date",
                    "required": false
                },
                "orderBy": {
                    "type": "enum",
                    "enum": [
                        "asc",
                        "desc"
                    ],
                    "default": "desc",
                    "required": false
                }
            }
        },
        {
            "methods": [
                "POST"
            ],
            "args": {
                "title": {
                    "type": "string",
                    "required": true
                },
                "shortDescription": {
                    "type": "string",
                    "required": false
                },
                "startDateTime": {
                    "type": "string",
                    "format": "date-time",
                    "required": false
                },
                "endDateTime": {
                    "type": "string",
                    "format": "date-time",
                    "required": false
                }
            }
        }
    ],
    "schema": {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "title": "campaign",
        "type": "object",
        "properties": {
            "id": {
                "type": "integer",
                "description": "Campaign ID."
            },
            "pageId": {
                "type": "integer",
                "description": "ID of the WordPress page assigned to the campaign."
            },
            "pagePermalink": {
                "type": "string",
                "format": "uri",
                "description": "Permalink for the campaign page."
            },
            "defaultFormId": {
                "type": "integer",
                "description": "ID of the default donation form for this campaign."
            },
            "defaultFormTitle": {
                "type": "string",
                "description": "Title of the default donation form."
            },
            "type": {
                "type": "string",
                "description": "Campaign type, e.g. &quot;core&quot;.",
                "enum": [
                    "core",
                    "p2p"
                ]
            },
            "title": {
                "type": "string",
                "description": "Campaign title.",
                "minLength": 3,
                "maxLength": 128,
                "errorMessage": "Campaign title is required."
            },
            "shortDescription": {
                "type": "string",
                "description": "Campaign short description",
                "maxLength": 120
            },
            "longDescription": {
                "type": "string",
                "description": "Long description of the campaign."
            },
            "logo": {
                "type": "string",
                "description": "URL to the campaign logo image."
            },
            "image": {
                "type": "string",
                "description": "URL to the campaign featured image."
            },
            "primaryColor": {
                "type": "string",
                "description": "Primary color used in campaign design."
            },
            "secondaryColor": {
                "type": "string",
                "description": "Secondary color used in campaign design."
            },
            "goal": {
                "type": "number",
                "description": "Goal amount/metric for the campaign.",
                "errorMessage": "Must be a number."
            },
            "goalType": {
                "enum": [
                    "amount",
                    "donations",
                    "donors",
                    "amountFromSubscriptions",
                    "subscriptions",
                    "donorsFromSubscriptions"
                ],
                "description": "Campaign goal type"
            },
            "goalStats": {
                "type": "object",
                "description": "Statistics about goal progress.",
                "properties": {
                    "actual": {
                        "type": "number",
                        "description": "Actual value raised so far."
                    },
                    "actualFormatted": {
                        "type": "string",
                        "description": "Formatted actual amount."
                    },
                    "percentage": {
                        "type": "number",
                        "description": "Percentage of the goal completed."
                    },
                    "goal": {
                        "type": "number",
                        "description": "Goal amount."
                    },
                    "goalFormatted": {
                        "type": "string",
                        "description": "Formatted goal amount."
                    }
                }
            },
            "status": {
                "enum": [
                    "active",
                    "inactive",
                    "draft",
                    "pending",
                    "processing",
                    "failed",
                    "archived"
                ],
                "description": "Current status of the campaign."
            },
            "startDate": {
                "type": [
                    "string",
                    "null"
                ],
                "format": "date-time",
                "description": "Start date of the campaign."
            },
            "endDate": {
                "type": [
                    "string",
                    "null"
                ],
                "format": "date-time",
                "description": "End date of the campaign (if set)."
            },
            "createdAt": {
                "type": "string",
                "format": "date-time",
                "description": "Timestamp when the campaign was created."
            }
        },
        "required": [
            "id",
            "title",
            "goal",
            "goalType"
        ],
        "allOf": [
            {
                "if": {
                    "properties": {
                        "goalType": {
                            "const": "amount"
                        }
                    }
                },
                "then": {
                    "properties": {
                        "goal": {
                            "type": "number"
                        }
                    },
                    "errorMessage": {
                        "properties": {
                            "goal": "Goal amount must be greater than 0"
                        }
                    }
                }
            },
            {
                "if": {
                    "properties": {
                        "goalType": {
                            "const": "donations"
                        }
                    }
                },
                "then": {
                    "properties": {
                        "goal": {
                            "minimum": 1,
                            "type": "number"
                        }
                    },
                    "errorMessage": {
                        "properties": {
                            "goal": "Number of donations must be greater than 0"
                        }
                    }
                }
            },
            {
                "if": {
                    "properties": {
                        "goalType": {
                            "const": "donors"
                        }
                    }
                },
                "then": {
                    "properties": {
                        "goal": {
                            "minimum": 1,
                            "type": "number"
                        }
                    },
                    "errorMessage": {
                        "properties": {
                            "goal": "Number of donors must be greater than 0"
                        }
                    }
                }
            },
            {
                "if": {
                    "properties": {
                        "goalType": {
                            "const": "amountFromSubscriptions"
                        }
                    }
                },
                "then": {
                    "properties": {
                        "goal": {
                            "minimum": 1,
                            "type": "number"
                        }
                    },
                    "errorMessage": {
                        "properties": {
                            "goal": "Goal recurring amount must be greater than 0"
                        }
                    }
                }
            },
            {
                "if": {
                    "properties": {
                        "goalType": {
                            "const": "subscriptions"
                        }
                    }
                },
                "then": {
                    "properties": {
                        "goal": {
                            "minimum": 1,
                            "type": "number"
                        }
                    },
                    "errorMessage": {
                        "properties": {
                            "goal": "Number of recurring donations must be greater than 0"
                        }
                    }
                }
            },
            {
                "if": {
                    "properties": {
                        "goalType": {
                            "const": "donorsFromSubscriptions"
                        }
                    }
                },
                "then": {
                    "properties": {
                        "goal": {
                            "minimum": 1,
                            "type": "number"
                        }
                    },
                    "errorMessage": {
                        "properties": {
                            "goal": "Number of recurring donors must be greater than 0"
                        }
                    }
                }
            }
        ]
    },
    "_links": {
        "self": [
            {
                "href": "https://local.lndo.site/wp-json/givewp/v3/campaigns"
            }
        ]
    }
}
```

</details>

---

### Compare this to a core WordPress endpoint response

```shell
curl -X https://local.lndo.site/wp-json/wp/v2/font-collections
```

<details>

<summary>Show  JSON Response</summary>

```json
{
    "namespace": "wp/v2",
    "methods": [
        "GET"
    ],
    "endpoints": [
        {
            "methods": [
                "GET"
            ],
            "args": {
                "context": {
                    "description": "Scope under which the request is made; determines fields present in response.",
                    "type": "string",
                    "enum": [
                        "view",
                        "embed",
                        "edit"
                    ],
                    "default": "view",
                    "required": false
                },
                "page": {
                    "description": "Current page of the collection.",
                    "type": "integer",
                    "default": 1,
                    "minimum": 1,
                    "required": false
                },
                "per_page": {
                    "description": "Maximum number of items to be returned in result set.",
                    "type": "integer",
                    "default": 10,
                    "minimum": 1,
                    "maximum": 100,
                    "required": false
                }
            }
        }
    ],
    "schema": {
        "$schema": "http://json-schema.org/draft-04/schema#",
        "title": "font-collection",
        "type": "object",
        "properties": {
            "slug": {
                "description": "Unique identifier for the font collection.",
                "type": "string",
                "context": [
                    "view",
                    "edit",
                    "embed"
                ],
                "readonly": true
            },
            "name": {
                "description": "The name for the font collection.",
                "type": "string",
                "context": [
                    "view",
                    "edit",
                    "embed"
                ]
            },
            "description": {
                "description": "The description for the font collection.",
                "type": "string",
                "context": [
                    "view",
                    "edit",
                    "embed"
                ]
            },
            "font_families": {
                "description": "The font families for the font collection.",
                "type": "array",
                "context": [
                    "view",
                    "edit",
                    "embed"
                ]
            },
            "categories": {
                "description": "The categories for the font collection.",
                "type": "array",
                "context": [
                    "view",
                    "edit",
                    "embed"
                ]
            }
        }
    },
    "_links": {
        "self": [
            {
                "href": "https://local.lndo.site/wp-json/wp/v2/font-collections"
            }
        ]
    }
}
```

</details>

## Affects

<!-- Mention any existing functionality affected by this PR to help inform the reviewer(s). -->

This affects most of the `/wp-json/givewp/v3` endpoints, **besides forms**.

## Visuals

<!-- Include screenshots or video to better communicate your changes. -->

All **but the forms** V3 REST endpoints should now include a `schema` property when requested with an `OPTIONS` method:

<img width="1448" height="885" alt="image" src="https://github.com/user-attachments/assets/0e4519a5-4538-4a52-9a65-da22f97222b7" />


## Testing Instructions

<!-- Help others test your PR as efficiently as possible. -->

Send a `GET` request to `/wp-json/givewp/v3/` to get all available endpoints.

The endpoints listed here, **besides any with form in the name**, when queried with an `OPTIONS` method should now return a valid `schema` property with a valid JSON schema.

In addition to that, pasting the object from the `schema` property should validate in an online validator, e.g. https://jsonschema.dev/

## Pre-review Checklist

<!-- Complete tasks prior to requesting a review. Add to this list, but do not remove the base items. -->

-   [ ] Acceptance criteria satisfied and marked in related issue
-   [x] Relevant `@unreleased` tags included in DocBlocks
-   [ ] Includes unit tests
-   [ ] Reviewed by the designer (if follows a design)
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

JasonTheAdams commented on src/API/REST/V3/Routes/Donations/DonationController.php:
```diff
@@ -982,8 +1036,71 @@ public function get_item_schema(): array
+                    ],
+                ],
+            ],
+            'required' => [
```
The item schema is used internally by WP in creation and update requests for validation. When it validates a property it uses _only_ the schema for that property (not the whole item schema). That's what I mean when I say "each property is its own schema". So having a required element that references properties by name won't even be seen by the validation system, so it won't know those properties are required.

JasonTheAdams replied:
These are fun quirks completely unique to how WordPress uses and processes the schema in validation. 🤪

defunctl commented:
@jonwaldstein @JasonTheAdams okay this creates a few more problems:

It looks like the admin frontend uses Ajv to validate schemas, however because the `schema` key was never in the right place, [this code would never any schema to validate against](https://github.com/impress-org/givewp/blob/4.6.1/src/Views/Components/AdminDetailsPage/index.tsx#L70-L77), but _now_ it does.

**This introduces its own issues:**
a. The Ajv library > [v7.0 does not support `draft-04`](https://ajv.js.org/guide/schema-language.html#draft-04), only the `draft-07` or above. I have removed the `$schema` keys referencing that to make this work for now, noting that I now know that WordPress doesn't support `draft-07`.
b. [The readonly prop](https://core.trac.wordpress.org/ticket/56152) is also not valid JSON schema which WordPress uses, the valid property is `readOnly`. So this also causes Ajv validation to fail. I since removed those, but we can put them back depending on what we decide below.
c. Using the `'required' => true` within a single property is also not valid JSON, which specifies it should be an array, and as I learned, WordPress requires this when using `rest_get_endpoint_args_for_schema()` and not using custom `args` for inputs.
d. Because Ajv was never getting actual schemas to validate against, it was also not apparent that Ajv was missing extended formats (like `email` etc...) and didn't allow union types due to strict mode, which I fixed for now.


_So, we have a major conflict here..._

@glaubersilva it looks like you're the one that originally added [this Ajv code](https://github.com/impress-org/givewp/blame/4.6.1/src/Views/Components/AdminDetailsPage/index.tsx#L70-L77).

I think there's only one path forward here: **Stop using Ajv on the frontend to validate schemas** because:
1. [Ajv doesn't support `draft-04`](https://ajv.js.org/guide/schema-language.html#draft-04) and [WordPress does not support `draft-07`](https://developer.wordpress.org/news/2024/07/json-schema-in-wordpress/#wordpress-rest-api).
2. WordPress has many non-spec JSON properties, which would cause Ajv to fail validation (`readonly` instead of `readOnly`, `required => true` etc...) and cause WordPress REST endpoints to not function as expected.

Considering it wasn't technically working until now, I think removing it would not cause any problems, but I really want everyone's feedback of which direction we should take on this before proceeding any further.

I hope this all makes sense!

defunctl commented on src/API/REST/V3/Routes/Donations/DonationController.php:
```diff
@@ -796,24 +794,34 @@ public function get_item_schema(): array
                 'lastName' => [
                     'type' => 'string',
                     'description' => esc_html__('Donor last name', 'give'),
-                    'format' => 'text-field',
```
Also, the help doc you provided doesn't list `text-field` as a valid format WordPress supports either, so I'm just a bit confused there as well.

<img width="974" height="332" alt="image" src="https://github.com/user-attachments/assets/165724bf-3106-4d4b-af8c-23a1c8d1d964" />

jonwaldstein replied:
@defunctl out of curiosity how were you able to notice the Ajv schema validation was failing?

FYI `text-field` was added in [WP 5.9.0 ](https://github.com/WordPress/WordPress/blob/86b7fc84bdff5a9db05a120bba7858d5f4d32364/wp-includes/rest-api.php#L2756) (I don't the docs are up to date) but you can see in the codebase

https://github.com/WordPress/WordPress/blob/86b7fc84bdff5a9db05a120bba7858d5f4d32364/wp-includes/rest-api.php#L2900

defunctl replied:
@jonwaldstein I noticed it failing in the dashboard with the console open: 

GiveWP > Donors > Click `Add note` which fires all that off and you'd get console errors.

<img width="1419" height="726" alt="image" src="https://github.com/user-attachments/assets/12cd6365-6f12-4370-a907-d43d2093a31f" />

defunctl replied:
If I checkout a commit much earlier before I removed the `readonly` stuff for example as well:

<img width="1419" height="726" alt="image" src="https://github.com/user-attachments/assets/30683156-a6bf-43d3-b6c5-a4fb4bd2f655" />
